### DISCUSSION TITLE
About the code

### DISCUSSION THREAD

**Discussion by yonghuming** [lawsie/guizero#504] [Ideas]
The code of guizero is elegant. I have read the source and found it easy to understand. I think i should write something to help others people have better understandings about guizero.

Most classe of guizero have less than 200 code excpet base.py or uitls.py and so on.
tkminxns seems very long, but it's really easy to undersand. Mixin seems an advanced language feature of python , maybe not common to use, you can treate minx as buff of player in MMORPG games. Each maxin just do one thing or control one property of class. 
```
class FocusMixin():
    def focus(self):
        """Give focus to the widget."""
        self.tk.focus_set()
```
This maxin just add focus method to any class that use or inherite it.
most property of guizero class are controlled using getter and setter funciton, so class code may seems longer than necessary.
When contributing to guizero, you should better keep  the mixins in in mind and sue mixins decrator classes.
guizero made a display object tree, just like display object list of flash player. We need class diagram in documents, this will help other people get involved easier.

I have some doubt. whe Specify click event of button, guizero ui command param of PushButton constructor, but I think `gpiozero` make a better work:
```
btn = Button(12)
def do_something():
    pass
btn.when_clicked = do_something
```
It's more like nature language and meaningful.
```
def do_somthing():
    pass

btn = PushButton(text='hello')
btn.when_clicked = do_something

message = Text(text='btn text')
message.source = button # set button 's value to Text 
```

Also others prefer config all widgets use option dict, in React or Vue, they use dict config almost everything.
We have event trigger so maybe we can assign a source widget for ohter.

Also maybe we can create a `guizero designer` just page for tk/tcl and qt desinger for qt.

![image](https://user-images.githubusercontent.com/435079/108981871-23bc1e80-76c8-11eb-99e0-804aee70f0a7.png)

I just draw a simple class diagram.

**Comment by martinohanlon** [comment]
Thank you for your suggestions. I will think about how we can include some additional information in the documentation for contributors.

There is a class diagram for guizero - you can find it here - https://github.com/lawsie/guizero/blob/master/docs-src/docs/images/guizero-class-hierarchy.jpg

It isnt include in the documentation, mainly because I didnt have a place to put it.

**Comment by yasenstar** [comment]
> Thank you for your suggestions. I will think about how we can include some additional information in the documentation for contributors.
> 
> There is a class diagram for guizero - you can find it here - https://github.com/lawsie/guizero/blob/master/docs-src/docs/images/guizero-class-hierarchy.jpg
> 
> It isnt include in the documentation, mainly because I didnt have a place to put it.

Hi @martinohanlon , the JPG file link you mentioned is not valid, could you check where is that file now? Thanks.

Xiaoqi

**Comment by yonghuming** [comment]
您好，邮件已经收到

**Comment by bsimmo** [comment]
Go to
https://github.com/lawsie/guizero/tree/master/docs-src/docs/images
Then look for it down the list.

**Comment by yasenstar** [comment]
Thanks, I can see the images on class-hierarchy.
